### PR TITLE
SI-8941 Idempotent presentation compilation of implicit classes

### PR DESCRIPTION
**Pull request by retronym**
When we name an implicit class, `enterImplicitWrapper` is called,
which enters the symbol for the factory method into the
owning scope. The tree defining this factory method is stowed into
`unit.synthetics`, from whence it will be retrieved and incorporated
into the enclosing tree during typechecking (`addDerivedTrees`).
The entry in `unit.synthetics` is removed at that point.

However, in the presentation compiler, we can typecheck a unit
more than once in a single run. For example, if, as happens
in the enclosed test, a call to ask for a type at a given
position interrupts type checking of the entire unit, we
can get into a situation whereby the first type checking
invocation has consumed the entry from `unit.synthetics`,
and the second will crash when it can't find an entry.

Similar problems have been solved in the past in
`enterExistingSym` in the presentation compiler. This method
is called when the namer encounters a tree that already has
a symbol attached. See 0b78a0196 / 148736c3df.

This commit takes a two pronged approach.

First, `enterExistingSym` is extended to handle implicit classes.
Any previous factory method in the owning scope is removed, and
`enterImplicitWrapper` is called to place a new tree for the factory
into `unit.synthetics` and to enter its symbol into the owning scope.

Second, the assertions that could be tripped in `addDerivedTrees`
and in `ImplicitClassWrapper#derivedSym` have been converted to
positioned errors.

The first change is sufficient to fix this bug, but the second
is also enough to make the enclosed test pass, and has been retained
as an extra layer of defence.

Review by @lrytz / @dragos

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala [hunk 1]
```diff
@@ -355,8 +357,9 @@ trait MethodSynthesis {
       def derivedSym: Symbol = {
         // Only methods will do! Don't want to pick up any stray
         // companion objects of the same name.
-        val result = enclClass.info decl name suchThat (x => x.isMethod && x.isSynthetic)
-        assert(result != NoSymbol, "not found: "+name+" in "+enclClass+" "+enclClass.info.decls)
+        val result = enclClass.info decl name filter (x => x.isMethod && x.isSynthetic)
+        if (result == NoSymbol && result.isOverloaded)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala]: should it be `||` here instead of `&&`?

**Comment by lrytz** [comment]
very nice work, great commit message!

File: src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala]: good catch, fixed

**Comment by retronym** [comment]
I've just pushed another commit that tests this, and other forms of idempotency, in a single threaded manner.

Incidentally, this reminds me of SI-5265. Idempotency can even be needed in the batch compiler, to make typechecking re-entract: https://github.com/retronym/scala/compare/ticket/5265

**Comment by lrytz** [comment]
LGTM!

**Pull request merged**